### PR TITLE
[OptRed] Extend -tritonintelgpu-optimize-reduction-locality to support `repCluster[0] > 2`

### PR DESCRIPTION
**Pull request by victor-eds**
Support `repCluster[0] > 2` by using 7-D tensors and adding a `convert_layout` operation before the final `reshape`.

See code for implementation details.

Signed-off-by: victor-eds <victor.perez@codeplay.com>

### CODE REVIEW

**Comment by victor-eds** [comment]
Part of https://github.com/intel/intel-xpu-backend-for-triton/issues/2266.

File: third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp [hunk 1]
```diff
@@ -83,66 +83,89 @@ static Value createReshapeForReduction(PatternRewriter &rewriter, Location loc,
...
+  ///         |                  | t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
+  ///         |                  v t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
+  ///         |                    ..................................................................................|
+  ///         |                    t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn | warpsPerCTA[2]
+  ///         |                    t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
+  /// size[1] |                    t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
+  ///         |                    t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
+  ///         v                    t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
   /// ```
   /// So we can reduce on dimensions 4 and 2 to get to:
```
Comment (jopperm) [inline third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp:106]:
```suggestion
  /// So we can reduce on dimensions 6 and 4 to get to:
```

File: third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp [hunk 2]
```diff
@@ -83,66 +83,89 @@ static Value createReshapeForReduction(PatternRewriter &rewriter, Location loc,
...
-  ///                  | t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
-  ///                  | t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
-  ///                  | t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn | warpsPerCTA[0]
-  ///                  | t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
-  /// sizePerThread[0] | t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
-  ///                  | t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
-  ///                  v t0 t1 t2 t3 ... tn t0 t1 t2 t3 ... tn tn1 tn2 tn3 ... tnn tn1 tn2 tn3 tn4 ... tnn |
+  ///                                                            warpsPerCTA[5]
+  ///                              <------------------------------------------------------------------------------->
+  ///                                           size[4]
```
Comment (jopperm) [inline third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp:90]: Clarify `size` refers to the tensor's shape (all other identifiers are attributes in the layout).

File: third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp [hunk 3]
```diff
@@ -222,10 +255,14 @@ struct DpasOperandPattern final : OpRewritePattern<ReduceOp> {
...
-    constexpr size_t rank = 5;
+    constexpr size_t rank = 7;
     std::array<int64_t, rank> shape{
         // Y axis
-        oldShape[0],
+        oldEncoding.getExecutionSize(),
+        (oldEncoding.getRepeatCount() * oldEncoding.getRepCluster()[0]) /
+            oldEncoding.getExecutionSize(),
+        oldShape[0] /
+            (oldEncoding.getRepeatCount() * oldEncoding.getRepCluster()[0]),
```
Comment (jopperm) [inline third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp:265]: Add comments explaining the dim (similar to what you did for the split X-axis.

**Review by etiotto — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp [hunk 4]
```diff
@@ -83,66 +75,89 @@ static Value createReshapeForReduction(PatternRewriter &rewriter, Location loc,
...
+  ///                   ^ t0 ^
+  ///                   | t1 |
+  /// threadsPerWarp[0] | t2 | warpsPerCTA[2]
+  ///                   | t3 |
+  /// ```
+  /// Reshaping from this layout to the final state would not work, as we would
+  /// end up modifying the number of elements per work-item (not allowed in
+  /// `reshape`).
+  ///
+  /// In order to avoid that, we can just convert the layout to a sliced layout
```
Comment (jopperm) [inline third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp:141]: This layout conversion is handled by the new subgroup transpose codegen you added in the other PRs, correct?

File: third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp [hunk 5]
```diff
@@ -83,66 +75,89 @@ static Value createReshapeForReduction(PatternRewriter &rewriter, Location loc,
...
   ///                    <------------------>
   ///                   ^ t0 t0 t0 t0 ... t0 tn1 tn1 tn1 ... tn1 ^
   ///                   | t1 t1 t1 t1 ... t1 tn2 tn2 tn2 ... tn2 |
-  /// threadsPerWarp[0] | t2 t2 t2 t2 ... t2 tn3 tn3 tn3 ... tn3 | warpsPerCTA[0]
+  /// threadsPerWarp[0] | t2 t2 t2 t2 ... t2 tn3 tn3 tn3 ... tn3 | warpsPerCTA[2]
   ///                   | t3 t3 t3 t3 ... t3 tn4 tn4 tn4 ... tn4 |
   /// ```
-  /// And on with step 3, reducing on dimension 1 and converting the layout to
-  /// the original one leads to the same output as the original operation.
+  /// And on with step 3, after reducing on dimension 3, we'd get:
```
Comment (jopperm) [inline third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp:130]: Could you annotate the dimension number in the figures, please?

**Review by jopperm — APPROVED** [review]
Some minor documentation nits, but LGTM otherwise!

File: third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp (hunk 4 above)
Comment (victor-eds) [reply third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp:141]: Yes!

File: third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp (hunk 5 above)
Comment (victor-eds) [reply third_party/intel/lib/TritonIntelGPUTransforms/OptimizeReductionLocality.cpp:130]: Not sure I understood this well. I added more information about the expected layout. As you can see, it's the same as the one above, but with a sliced dimension (3).

**Pull request merged**